### PR TITLE
adding weights to positive examples

### LINKED ISSUE #1744: Improve binary focal loss to trade off recall and precision by adding weights to positive examples like torch.nn.BCEWithLogitsLoss
## 🚀 Feature
Improve binary focal loss to trade off recall and precision by adding weights to positive examples. Just like `pos_weight` parameter in torch.nn.BCEWithLogitsLoss. `pos_weight` must be a vector with length equal to the number of classes.

### PR DESCRIPTION
**Pull request by MrShevan**
#### Changes

Fixes #1744

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/losses/focal.py [hunk 1]
```diff
@@ -207,9 +211,17 @@ def binary_focal_loss_with_logits(
     if input.size(0) != target.size(0):
         raise ValueError(f'Expected input batch_size ({input.size(0)}) to match target batch_size ({target.size(0)}).')
 
+    if pos_weight is None:
+        pos_weight = torch.ones(input.size(-1), device=input.device)
```
Comment (edgarriba) [inline kornia/losses/focal.py:215]:
```suggestion
        pos_weight = torch.ones(input.size(-1), device=input.device, dtype=input.dtype)
```
we should infer also dtypes

File: kornia/losses/focal.py [hunk 2]
```diff
@@ -207,9 +211,17 @@ def binary_focal_loss_with_logits(
...
+    elif not isinstance(pos_weight, torch.Tensor):
```
Comment (edgarriba) [inline kornia/losses/focal.py:216]: i suggest to use the new testing api here: https://github.com/kornia/kornia/blob/master/kornia/testing/__init__.py#L208-L228

and eventually update the existing checks

Comment (MrShevan) [reply kornia/losses/focal.py:216]: I did it for checking 'is tensor', but when I try to replace condition of checking 'equals to number of classes' with 
`KORNIA_CHECK(input.size(-1)` == pos_weight.size(0), "Expected pos_weight equals number of classes.")`

it's problem for jit test:
`Expected a value of type 'Tensor (inferred)' for argument 'condition' but instead found type 'bool'. Inferred 'condition' to be of type 'Tensor' because it was not annotated with an explicit type.`

Comment (MrShevan) [reply kornia/losses/focal.py:216]: I resolved this problem by adding typo for `condition` param:
`def KORNIA_CHECK(condition: bool, msg: Optional[str] = None):`

Comment (MrShevan) [reply kornia/losses/focal.py]: @edgarriba Is this a valid solution?

Comment (edgarriba) [reply kornia/losses/focal.py]: I think yes as soon as all test pass. Torchscript is highly sensible and tricky

Comment (MrShevan) [reply kornia/losses/focal.py]: There was one fail in running test adjust.py. 

I changed `KORNIA_CHECK((factor >= 0).any(), f"Contrast factor must be positive. Got {factor}")` to `KORNIA_CHECK(any(factor >= 0), f"Contrast factor must be positive. Got {factor}")`. Python function any() outputs bool value, but output of .any() not declared.

Can you approve running tests?

File: kornia/losses/focal.py [hunk 3]
```diff
@@ -198,18 +204,25 @@ def binary_focal_loss_with_logits(
...
+        raise TypeError(f"Input type is not a Tensor. Got {type(input)}")
 
     if not len(input.shape) >= 2:
         raise ValueError(f"Invalid input shape, we expect BxCx*. Got: {input.shape}")
 
     if input.size(0) != target.size(0):
         raise ValueError(f'Expected input batch_size ({input.size(0)}) to match target batch_size ({target.size(0)}).')
 
+    if pos_weight is None:
+        pos_weight = torch.ones(input.size(-1), device=input.device, dtype=input.dtype)
```
Comment (xiaoyuan0203) [inline kornia/losses/focal.py:217]: I suggest a modification of `input.size(-1)` because of a pytorch-style input size with

image

in [torch.nn.CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss), the same as other places like `pos_weight` size check.

Can also add a doc to describe the input size.

Comment (MrShevan) [reply kornia/losses/focal.py:217]: @xiaoyuan0203 You are right. I added you as collaborator in my forked branch (I couldn't push my changes without it into your forked repo), so you can make your changes to `input.size(-1)`

File: kornia/losses/focal.py (hunk 2 above)
Comment (edgarriba) [reply kornia/losses/focal.py]: i checked that one from my side and seems fine :)

File: kornia/losses/focal.py [hunk 4]
```diff
@@ -60,8 +62,8 @@ def focal_loss(
             stacklevel=2,
         )
 
-    if not isinstance(input, torch.Tensor):
-        raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
+    if not isinstance(input, Tensor):
+        raise TypeError(f"Input type is not a Tensor. Got {type(input)}")
```
Comment (edgarriba) [inline kornia/losses/focal.py:66]: `KORNIA_CHECK_SHAPE(input, ["B", "C", "*"])` which evnetually check for tensor too

Comment (edgarriba) [reply kornia/losses/focal.py:66]: add also one too for the target similar to this one

File: kornia/losses/focal.py [hunk 5]
```diff
@@ -78,11 +80,11 @@ def focal_loss(
         raise ValueError(f"input and target must be in the same device. Got: {input.device} and {target.device}")
```
Comment (edgarriba) [inline kornia/losses/focal.py:80]: `KORNIA_CHECK(input.device == target.device, f"input and target must be in the same device. Got: {input.device} and {target.device}")`

File: kornia/losses/focal.py [hunk 6]
```diff
@@ -78,11 +80,11 @@ def focal_loss(
...
 
     # compute softmax over the classes axis
-    input_soft: torch.Tensor = F.softmax(input, dim=1)
-    log_input_soft: torch.Tensor = F.log_softmax(input, dim=1)
+    input_soft: Tensor = F.softmax(input, dim=1)
```
Comment (edgarriba) [inline kornia/losses/focal.py:83]:
```suggestion
    input_soft: Tensor = input.softmax(1)
```
i'd suggest to avoid torch imports so that its easy later to extend to numpy/jax

File: kornia/losses/focal.py [hunk 7]
```diff
@@ -198,18 +204,25 @@ def binary_focal_loss_with_logits(
             stacklevel=2,
         )
 
-    if not isinstance(input, torch.Tensor):
-        raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
+    if not isinstance(input, Tensor):
```
Comment (edgarriba) [inline kornia/losses/focal.py:207]: use KORNIA_CHECK api here too

File: kornia/losses/focal.py [hunk 8]
```diff
@@ -198,18 +204,25 @@ def binary_focal_loss_with_logits(
...
         raise ValueError(f"Invalid input shape, we expect BxCx*. Got: {input.shape}")
 
     if input.size(0) != target.size(0):
         raise ValueError(f'Expected input batch_size ({input.size(0)}) to match target batch_size ({target.size(0)}).')
 
+    if pos_weight is None:
+        pos_weight = torch.ones(input.size(-1), device=input.device, dtype=input.dtype)
+
+    KORNIA_CHECK_IS_TENSOR(pos_weight)
+    KORNIA_CHECK(input.size(-1) == pos_weight.size(0), "Expected pos_weight equals number of "
```
Comment (edgarriba) [inline kornia/losses/focal.py:220]:
```suggestion
    KORNIA_CHECK(input.shape[-1] == pos_weight.shape[0], "Expected pos_weight equals number of "
```
more numpy/jax style

**Review by edgarriba — APPROVED** [review]
LGTM. Added some suggestions to make the code more python agnostic

File: kornia/losses/focal.py [hunk 9]
```diff
@@ -198,18 +204,25 @@ def binary_focal_loss_with_logits(
...
+                                                       "classes.")
+
     probs_pos = torch.sigmoid(input)
     probs_neg = torch.sigmoid(-input)
-    loss_tmp = -alpha * torch.pow(probs_neg, gamma) * target * F.logsigmoid(input) - (
+    loss_tmp = -alpha * pos_weight * torch.pow(probs_neg, gamma) * target * F.logsigmoid(input) - (
```
[the remark this answers is not in the document]
Comment (MrShevan) [reply kornia/losses/focal.py]: I replaced `F.log_softmax(input, dim=1)` with `input.log_softmax(1)`

Unfortunately Tensor doesn't exist method logsigmoid(), so to avoid using nn.functional, I replaced it with `input.sigmoid().log()` which the same and `-input.sigmoid().log()`

[the remark this answers is not in the document]
Comment (MrShevan) [reply kornia/losses/focal.py]: So finally I deleted `import torch.nn.functional as F`

File: kornia/losses/focal.py (hunk 3 above)
Comment (MrShevan) [reply kornia/losses/focal.py]: I think that it is advisable to use here only general constructions from PyTorch and use more numpy/jax style. 

What you say, may be changed here:
[https://github.com/facebookresearch/fvcore/blob/main/fvcore/nn/focal_loss.py#L36](https://github.com/facebookresearch/fvcore/blob/main/fvcore/nn/focal_loss.py#L36)

The `F.binary_cross_entropy_with_logits(inputs, targets, reduction="none")` is already used [here](https://github.com/facebookresearch/fvcore/blob/main/fvcore/nn/focal_loss.py#L36), so it can be easy add pos_weight as argument.

File: kornia/losses/focal.py [hunk 10]
```diff
@@ -178,14 +174,17 @@ def binary_focal_loss_with_logits(
...
+          Must be a vector with length equal to the number of classes.
 
     Returns:
         the computed loss.
 
     Examples:
         >>> kwargs = {"alpha": 0.25, "gamma": 2.0, "reduction": 'mean'}
-        >>> logits = torch.tensor([[[6.325]],[[5.26]],[[87.49]]])
-        >>> labels = torch.tensor([[[1.]],[[1.]],[[0.]]])
+        >>> logits = tensor([[[6.325]],[[5.26]],[[87.49]]])
```
Comment (edgarriba) [inline kornia/losses/focal.py:186]:
```suggestion
        >>> logits = torch.tensor([[[6.325]],[[5.26]],[[87.49]]])
```

File: kornia/losses/focal.py [hunk 11]
```diff
@@ -178,14 +174,17 @@ def binary_focal_loss_with_logits(
...
+        >>> labels = tensor([[[1.]],[[1.]],[[0.]]])
```
Comment (edgarriba) [inline kornia/losses/focal.py:187]:
```suggestion
        >>> labels = torch.tensor([[[1.]],[[1.]],[[0.]]])
```

**Comment by MrShevan** [comment]
@edgarriba I Approved suggestions for doctests :) Could you run all tests again?

**Comment by MrShevan** [comment]
There were my mistake, I replaced `F.logsigmoid(-input)` with `-input.sigmoid().log()`, but it was right to do so `(-input).sigmoid().log()`
